sign/ed448/modular.go

```diff
@@ -0,0 +1,196 @@
+package ed448
+
+import (
+	"encoding/binary"
+	"math/bits"
+)
+
+// order is 2^446-0x8335dc163bb124b65129c96fde933d8d723a70aadc873d6d54a7bb0d,
+// which is the number of points in the prime subgroup.
+var order = [Size]byte{
+	0xf3, 0x44, 0x58, 0xab, 0x92, 0xc2, 0x78, 0x23,
+	0x55, 0x8f, 0xc5, 0x8d, 0x72, 0xc2, 0x6c, 0x21,
+	0x90, 0x36, 0xd6, 0xae, 0x49, 0xdb, 0x4e, 0xc4,
+	0xe9, 0x23, 0xca, 0x7c, 0xff, 0xff, 0xff, 0xff,
+	0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
+	0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
+	0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0x3f,
+	0x00,
+}
+
+// residue446 is 2^446 mod order.
+var residue446 = [32]byte{
+	0x0d, 0xbb, 0xa7, 0x54, 0x6d, 0x3d, 0x87, 0xdc,
+	0xaa, 0x70, 0x3a, 0x72, 0x8d, 0x3d, 0x93, 0xde,
+	0x6f, 0xc9, 0x29, 0x51, 0xb6, 0x24, 0xb1, 0x3b,
+	0x16, 0xdc, 0x35, 0x83, 0x00, 0x00, 0x00, 0x00,
+}
+
+// residue448 is 2^448 mod order.
+var residue448 = [32]byte{
+	0x34, 0xec, 0x9e, 0x52, 0xb5, 0xf5, 0x1c, 0x72,
+	0xab, 0xc2, 0xe9, 0xc8, 0x35, 0xf6, 0x4c, 0x7a,
+	0xbf, 0x25, 0xa7, 0x44, 0xd9, 0x92, 0xc4, 0xee,
+	0x58, 0x70, 0xd7, 0x0c, 0x02, 0x00, 0x00, 0x00,
+}
+
+// invFour is 1/4 mod order.
+var invFour = [Size]byte{
+	0x3d, 0x11, 0xd6, 0xaa, 0xa4, 0x30, 0xde, 0x48,
+	0xd5, 0x63, 0x71, 0xa3, 0x9c, 0x30, 0x5b, 0x08,
+	0xa4, 0x8d, 0xb5, 0x6b, 0xd2, 0xb6, 0x13, 0x71,
+	0xfa, 0x88, 0x32, 0xdf, 0xff, 0xff, 0xff, 0xff,
+	0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
+	0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
+	0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0x0f,
+}
+
+// isLessThan returns true if 0 <= x < y, and assumes that slices have the same length.
+func isLessThan(x, y []byte) bool {
+	i := len(x) - 1
+	for i > 0 && x[i] == y[i] {
+		i--
+	}
+	return x[i] < y[i]
+}
+
+func byte2uint(z []uint, x []byte) {
```

Comment:
Might be worthwhile to  either
1. create specialisations for the actual sizes used in `reduce` or
2. use unsafe.Pointer to simply reinterpret one as the other.

Reply:
(Or well... there is more to gain by speeding up multiplication of scalars.)

Reply:
I will use unsafe pointer conversion. nice suggestion